webknossos/script-collection/globalize_floodfill.py

```diff
@@ -0,0 +1,364 @@
+    a: Tuple[int, int, int], b: Tuple[int, int, int]
+) -> Tuple[int, int, int]:
+    return a[0] + b[0], a[1] + b[1], a[2] + b[2]
+
+
+def substract_positions(
+    a: Tuple[int, int, int], b: Tuple[int, int, int]
+) -> Tuple[int, int, int]:
+    return a[0] - b[0], a[1] - b[1], a[2] - b[2]
+
+
+def inside_bbox(
+    point: Tuple[int, int, int],
+    bbox_top_left: Tuple[int, int, int] = (0, 0, 0),
+    bbox_bottom_right: Tuple[int, int, int] = CUBE_SHAPE,
+) -> bool:
+    return (
+        bbox_top_left[0] <= point[0] < bbox_bottom_right[0]
+        and bbox_top_left[1] <= point[1] < bbox_bottom_right[1]
+        and bbox_top_left[2] <= point[2] < bbox_bottom_right[2]
+    )
+
+
+def execute_floodfill(
+    data_path: Path,
+    seed_position: Tuple[int, int, int],
+    already_processed_bbox: BoundingBox,
+    source_id: int,
+    target_id: int,
+) -> None:
+    # bucketData[neighbourVoxelIndex] == sourceCellId || (isInsideBBox & & bucketData[neighbourVoxelIndex] == targetCellId)
+    bucket_and_seed_pos: List[Tuple[Tuple[int, int, int], Tuple[int, int, int]]] = [
+        get_bucket_pos_and_offset(seed_position)
+    ]
+    # bucket_to_seed_pos: Dict[
+    #    Tuple[int, int, int], Set[Tuple[int, int, int]]
+    # ] = defaultdict(set)
+    already_processed_bbox_top_left = already_processed_bbox.topleft.to_tuple()
+    already_processed_bbox_bottom_right = already_processed_bbox.bottomright.to_tuple()
+    visited = np.zeros(
+        already_processed_bbox.size.to_tuple(), dtype=np.uint8
+    )  # bitarray needs less memory, but new dependency
+    bucket_count = 0
+    with wkw.Dataset.open(str(data_path)) as wkw_data:
+        while len(bucket_and_seed_pos) != 0:
+            bucket_count += 1
+            if bucket_count % 100 == 0:
+                print("Handled seed positions ", bucket_count)
+
+            dirty_bucket = False
+            current_cube, seed_position = bucket_and_seed_pos.pop()
+            global_seed_position = add_positions(current_cube, seed_position)
+            cube_data = wkw_data.read(current_cube, CUBE_SHAPE)
+            cube_data = cube_data[0, :, :, :]
+
+            if cube_data[seed_position] == source_id or (
+                inside_bbox(
+                    global_seed_position,
+                    already_processed_bbox_top_left,
+                    already_processed_bbox_bottom_right,
+                )
+                and cube_data[seed_position] == target_id
+                and not visited[
+                    substract_positions(
+                        global_seed_position, already_processed_bbox_top_left
+                    )
+                ]
+            ):
+                seeds_in_curr_bucket: Set[Tuple[int, int, int]] = set()
+                seeds_in_curr_bucket.add(seed_position)
+                while len(seeds_in_curr_bucket) > 0:
+                    seed_pos = seeds_in_curr_bucket.pop()
+                    global_seed_pos = add_positions(current_cube, seed_pos)
```

Comment:
there's `seed_pos` and `seed_position` as well as `global_seed_pos` and `global_seed_position`. Either, these variables should overwrite the others, or the naming should be clearer to distinguish between the two semantics.